add NAF filter for clinics

tingaloo opened the pull request:
I rule and have completed some work on Case Manager that's ready for review!

Adds a checkbox that DISABLES clinics that DO NOT accept NAF funding.

This pull request makes the following changes:
* Adds checkbox to filter NAF

![screen shot 2017-04-08 at 7 38 26 pm](https://cloud.githubusercontent.com/assets/8662824/24833311/f93fdb94-1c92-11e7-82e1-7341a9ecbd9f.jpg)


Clinics that do not define accepts_naf will be sorted as false automatically, in this case 'Sample Clinic 1' and 'Sample Clinic 2'. 'SAMPLE CLINIC without NAF' is disabled but still selected.
<br>


I made this more Ruby heavy than JS, doing all lookups and array manipulating in the patient helper. For now I disabled clinics because it is easier and I have no idea how many clinics are in production. If we want to remove them, I think we'd have to override some bootstrap styles because option tags.

Currently works in development, and should work with any array passed in.

It relates to the following issue #s: 
* Fixes #979

colinxfleming commented:
MY DUDE. Thank you! I'm running around this evening and will review
tomorrow or monday.

On Sat, Apr 8, 2017, 7:40 PM Lew <notifications@github.com> wrote:

> I rule and have completed some work on Case Manager that's ready for
> review!
>
> Adds a checkbox that DISABLES clinics that DO NOT accept NAF funding.
>
> This pull request makes the following changes:
>
>    - Adds checkbox to filter NAF
>
> [image: screen shot 2017-04-08 at 7 38 26 pm]
> <https://cloud.githubusercontent.com/assets/8662824/24833311/f93fdb94-1c92-11e7-82e1-7341a9ecbd9f.jpg>
>
> Clinics that do not define accepts_naf will be sorted as false
> automatically, in this case 'Sample Clinic 1' and 'Sample Clinic 2'.
> 'SAMPLE CLINIC without NAF' is disabled but still selected.
>
> I made this more Ruby heavy than JS, doing all lookups and array
> manipulating in the patient helper. For now I disabled clinics because it
> is easier and I have no idea how many clinics are in production. If we want
> to remove them, I think we'd have to override some bootstrap styles because
> option tags.
>
> Currently works in development, and should work with any array passed in.
>
> It relates to the following issue #s:
>
>    - Fixes #979
>    <https://github.com/colinxfleming/dcaf_case_management/issues/979>
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/colinxfleming/dcaf_case_management/pull/986
> Commit Summary
>
>    - added NAF filter for clinics
>    - minor styling
>
> File Changes
>
>    - *M* app/assets/javascripts/patients.coffee
>    <https://github.com/colinxfleming/dcaf_case_management/pull/986/files#diff-0>
>    (8)
>    - *M* app/helpers/clinics_helper.rb
>    <https://github.com/colinxfleming/dcaf_case_management/pull/986/files#diff-1>
>    (4)
>    - *M* app/helpers/patients_helper.rb
>    <https://github.com/colinxfleming/dcaf_case_management/pull/986/files#diff-2>
>    (9)
>    - *M* app/views/patients/_abortion_information.html.erb
>    <https://github.com/colinxfleming/dcaf_case_management/pull/986/files#diff-3>
>    (3)
>    - *M* db/seeds.rb
>    <https://github.com/colinxfleming/dcaf_case_management/pull/986/files#diff-4>
>    (15)
>
> Patch Links:
>
>    - https://github.com/colinxfleming/dcaf_case_management/pull/986.patch
>    - https://github.com/colinxfleming/dcaf_case_management/pull/986.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/colinxfleming/dcaf_case_management/pull/986>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADsA9GBe0FFO0d6_1e30X4Sq_a3PxXesks5ruBr0gaJpZM4M33cE>
> .
>

colinxfleming commented:
This PR is a beaut. I think we can merge as is, I'll plan on getting it in this week. Thanks a ton for knocking this out so quickly.

colinxfleming commented:
whoops, I can't quite get CI to pass. I suspect it has something to do with the JS causing some kind of redirect -- I would **guess** that moving it away from `document` to something else will resolve. Not sure what that something else is off the top of my head but I'm sure there's an example you can crib off.

This test is probably the canary and I suspect when you get this to pass everything else will fall into place:

```
Expected response to be a <success>, but was <302>
test/controllers/patients_controller_test.rb:63:in `block (2 levels) in <class:PatientsControllerTest>'
```

Sorry for the false flag! (It does look good!)

tingaloo commented:
Actually not a false flag, and not a JS problem either.

This line fails with good reason (not sure why it didn't fail before).
```clinic_with_naf = array.map { |clinics| [clinics, :"data-naf" => Clinic.find_by(name: clinics).accepts_naf] }```

```Clinic.find_by``` currently doesn't have enough time to fetch more than 2 clinics. So it kind of looks like this.

```
MONGODB | localhost:27017 | dcaf_case_management_development.find | STARTED | {"find"=>"clinics", "filter"=>{"name"=>"Sample Clinic 1 - DC"}}
MONGODB | localhost:27017 | dcaf_case_management_development.find | SUCCEEDED | 0.001824s
#<Clinic:0x007fc02bf702b8>
MONGODB | localhost:27017 | dcaf_case_management_development.find | STARTED | {"find"=>"clinics", "filter"=>{"name"=>"Sample Clinic 2 - VA"}}
MONGODB | localhost:27017 | dcaf_case_management_development.find | SUCCEEDED | 0.001461s
#<Clinic:0x007fc0320d3870>
MONGODB | localhost:27017 | dcaf_case_management_development.find | STARTED | {"find"=>"clinics", "filter"=>{"name"=>"Sample Clinic with NAF"}}
MONGODB | localhost:27017 | dcaf_case_management_development.find | SUCCEEDED | 0.00132s
  Rendered patients/_abortion_information.html.erb (15.2ms)
  Rendered patients/edit.html.erb within layouts/application (111.6ms)
Redirected to http://localhost:3000/
Completed 302 Found in 178ms
```
Basically if there's more than 2 clinics, the page will load before Clinic.find_by gets a chance to fetch. I'll rewrite this in JS since it's frontend and async.

colinxfleming commented:
Ah, i see. We are currently in the process of moving away from a list of
clinics as an env var to a proper clinic object system. There's an open pr
moving over the rest ofthe way that might be helpful to refer to.

On Mon, Apr 10, 2017, 9:04 PM Lew <notifications@github.com> wrote:

> Actually not a false flag, and not a JS problem either.
>
> This line fails with good reason (not sure why it didn't fail before).
> clinic_with_naf = array.map { |clinics| [clinics, :"data-naf" =>
> Clinic.find_by(name: clinics).accepts_naf] }
>
> Clinic.find_by currently doesn't have enough time to fetch more than 2
> clinics. So it kind of looks like this.
>
> MONGODB | localhost:27017 | dcaf_case_management_development.find | STARTED | {"find"=>"clinics", "filter"=>{"name"=>"Sample Clinic 1 - DC"}}
> MONGODB | localhost:27017 | dcaf_case_management_development.find | SUCCEEDED | 0.001824s
> #<Clinic:0x007fc02bf702b8>
> MONGODB | localhost:27017 | dcaf_case_management_development.find | STARTED | {"find"=>"clinics", "filter"=>{"name"=>"Sample Clinic 2 - VA"}}
> MONGODB | localhost:27017 | dcaf_case_management_development.find | SUCCEEDED | 0.001461s
> #<Clinic:0x007fc0320d3870>
> MONGODB | localhost:27017 | dcaf_case_management_development.find | STARTED | {"find"=>"clinics", "filter"=>{"name"=>"Sample Clinic with NAF"}}
> MONGODB | localhost:27017 | dcaf_case_management_development.find | SUCCEEDED | 0.00132s
>   Rendered patients/_abortion_information.html.erb (15.2ms)
>   Rendered patients/edit.html.erb within layouts/application (111.6ms)
> Redirected to http://localhost:3000/
> Completed 302 Found in 178ms
>
> Basically if there's more than 2 clinics, the page will load before
> Clinic.find_by gets a chance to fetch. I'll rewrite this in JS since it's
> frontend and async.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/colinxfleming/dcaf_case_management/pull/986#issuecomment-293120438>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADsA9AruIAqryJ847JmtBBmyRixbvUYvks5rutGagaJpZM4M33cE>
> .
>

colinxfleming commented:
( #975 is that PR)

tingaloo commented:
Not really fond of jquery ajax and coffeescript, but I got it to work.

Patients javascript is quite bloated now. I added some routes for my JS calls because I'm not sure how to call helper methods using AJAX, so the code can be a bit janky.
 
I wrote around this line from #975  
```Clinic.where(active: true).map { |clinic| [clinic.name, clinic.id] }```
Overall it should be good to go.

colinxfleming commented on app/controllers/patients_controller.rb:
```diff
@@ -56,6 +56,17 @@ def data_entry_create # temporary
   end
+
+
+  def fetch_clinic_naf
```
What are your thoughts about moving this to the `clinics` controller? Can rig it up something like `ClinicsController#naf_status`?

colinxfleming commented on app/controllers/patients_controller.rb:
```diff
@@ -56,6 +56,17 @@ def data_entry_create # temporary
+
+
+  def fetch_clinic_naf
+    accepts_naf =  Clinic.find_by(id: params['id']).accepts_naf
```
You can just use `Clinic.find(params['id'])` here, no need to declare the field. A nice little rails-ism!

colinxfleming commented on app/controllers/patients_controller.rb:
```diff
@@ -56,6 +56,17 @@ def data_entry_create # temporary
+    render json: [params['id'], accepts_naf]
+  end
+
+  def get_clinics
```
Same deal -- lets move this to the clinics controller. Might make a good thing to return from the index route if you're requesting JSON?

colinxfleming commented on config/routes.rb:
```diff
@@ -10,6 +10,8 @@
       resources :notes, only: [ :create, :update ]
       resources :external_pledges, only: [ :create, :update, :destroy ]
     end
+    get 'get_clinics', to: 'patients#get_clinics', as: 'get_clinics'
```
Per above, I think these might be better suited for the clinics controller. What are your thoughts?

colinxfleming commented on app/assets/javascripts/patients.coffee:
```diff
@@ -22,6 +21,39 @@ calculateRemainder = ->
 valueToNumber = (val) ->
   +val || 0
 
+getClinics = ->
```
We have a clinics coffeescript file, no? Let's move stuff pertaining to this feature over there, since as you very correctly pointed out, the patients JS is getting a bit massive.

tingaloo commented:
All tidied up! Moving the logic to `Clinics` cleaned A LOT of the logic in `Patients`. Also fetching all the clinics at once via `index` made things much more simple. I got tunnel vision working on this the first time, now it looks way better.

Only thing that's left is an ajax integration test, I tried writing [a test](https://github.com/colinxfleming/dcaf_case_management/pull/986/commits/bcced15a14e9072854241f41fb1cb4fdea8487a4) but it kept failing. I'll give it another shot at the next hacknight.

Happy Easter!

colinxfleming commented on test/controllers/clinics_controller_test.rb:
```diff
@@ -21,6 +21,16 @@ class ClinicsControllerTest < ActionDispatch::IntegrationTest
   end
 
+  describe 'naf_status' do
+    before {
```
You can simplify this a bit -- `create` lets you pass in a hash of stuff as an argument, and it'll set the object attributes accordingly. I think something like `before { @clinic = create :clinic, accepts_naf: true }` should fly.

colinxfleming commented on config/routes.rb:
```diff
@@ -8,19 +8,22 @@
     end
 
+    get 'naf_status', to: 'clinics#naf_status', as: 'naf_status'
+    get 'clinics', to: 'clinics#index', as: 'clinics'
```
I think we already have this route set as part of `resources :clinics` -- I think we can just use that. (We should keep `naf_status` though for sure!)

colinxfleming commented on app/controllers/clinics_controller.rb:
```diff
@@ -5,6 +5,7 @@ class ClinicsController < ApplicationController
 
   def index
     @clinics = Clinic.all
+    render json: @clinics.to_json
```
To make it a bit clearer that this endpoint can be either json or HTML, let's throw this in a `respond_to` block. Something like:

```ruby
respond_to do |format|
  format.html
  format.json { render @clinics.to_json }
end
```

colinxfleming commented on config/routes.rb:
```diff
@@ -8,19 +8,22 @@
                 only: [ :create, :update, :destroy ]
     end
 
+    get 'naf_status', to: 'clinics#naf_status', as: 'naf_status'
```
Do we need this endpoint? Javascript is admittedly not my strong suit but it doesn't look like it's getting called anymore. Thoughts?

colinxfleming commented on test/controllers/clinics_controller_test.rb:
```diff
@@ -21,6 +21,16 @@ class ClinicsControllerTest < ActionDispatch::IntegrationTest
+      @clinic = create :clinic
+      @clinic[:accepts_naf] = true
+     }
+    it 'should render true' do
```
HMU if you have a tough time with this -- rails 5 changed how they did controller tests. http://guides.rubyonrails.org/testing.html#functional-tests-for-your-controllers is the documentation but I spent a lot of quality time with this recently and it's fresh in my mind.

colinxfleming commented on test/controllers/clinics_controller_test.rb:
```diff
@@ -21,6 +21,16 @@ class ClinicsControllerTest < ActionDispatch::IntegrationTest
     end
```
An additional test in this controller to confirm that it responds to `index` as json would be great! Just a ping to confirm that the endpoint returns JSON data roughly in the format you're consuming.

tingaloo commented:
I'd say it's done! Made changes and removed the naf_status function.

Removed the capybara test because the CSS attribute never returned anything useful. Maybe somebody with more capy exp can give it a go.
